zanata-war/src/main/java/org/zanata/service/impl/MergeTranslationsServiceImpl.java

```diff
@@ -0,0 +1,278 @@
+            return AsyncTaskResult.taskResult();
+        }
+
+        if (getSupportedLocales(targetProjectSlug, targetVersionSlug).isEmpty()) {
+            log.error("No locales enabled in target version of {} [{}]",
+                    targetProjectSlug, targetVersionSlug);
+            return AsyncTaskResult.taskResult();
+        }
+
+        Optional<MergeTranslationsTaskHandle> taskHandleOpt =
+                Optional.fromNullable(handle);
+
+        if (taskHandleOpt.isPresent()) {
+            prepareMergeTranslationsHandle(sourceVersion, targetVersion,
+                    taskHandleOpt.get());
+        }
+
+        Stopwatch overallStopwatch = Stopwatch.createStarted();
+        log.info("merge translations start: from {} to {}", sourceProjectSlug
+                + ":" + sourceVersionSlug, targetProjectSlug + ":"
+                + targetVersionSlug);
+
+        int startCount = 0;
+        int totalCount = getTotalMatchCount(sourceVersion.getId(),
+                targetVersion.getId());
+
+        List<HLocale> supportedLocales = getSupportedLocales(targetVersion
+                .getProject().getSlug(), targetVersion.getSlug());
+
+        while (startCount < totalCount) {
+            int processedCount =
+                    mergeTranslationBatch(sourceVersion, targetVersion,
+                            supportedLocales, useNewerTranslation, startCount,
+                            TRANSLATION_BATCH_SIZE);
+            if (taskHandleOpt.isPresent()) {
+                taskHandleOpt.get().increaseProgress(processedCount);
+            }
+
+            startCount += TRANSLATION_BATCH_SIZE;
+            textFlowDAO.clear();
+        }
+        versionStateCacheImpl.clearVersionStatsCache(targetVersion.getId());
+        log.info("merge translation end: from {} to {}, {}", sourceProjectSlug
+                + ":" + sourceVersionSlug, targetProjectSlug + ":"
+                + targetVersionSlug, overallStopwatch);
+
+        return AsyncTaskResult.taskResult();
+    }
+
+    protected int mergeTranslationBatch(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion, List<HLocale> supportedLocales,
+            boolean useNewerTranslation, int offset, int batchSize) {
+        try {
+            return new MergeTranslationsWork(sourceVersion.getId(),
+                    targetVersion.getId(), offset, batchSize,
+                    useNewerTranslation, supportedLocales,
+                    textFlowDAO, translationStateCacheImpl)
+                    .workInTransaction();
+        } catch (Exception e) {
+            log.warn("exception during copy text flow target", e);
+            return 0;
+        }
+    }
+
+    /**
+     * Check if sourceVersion or targetVersion has source document.
+     *
+     * @param sourceVersion
+     * @param targetVersion
+     */
+    private boolean isVersionsEmpty(HProjectIteration sourceVersion,
+            HProjectIteration targetVersion) {
```

Comment:
For our production servers, the dropdown would have a lot of projects so it would be nice to have it able to be filtered by typing the name at least, but it is not an everyday thing to do so I think we should only try to do something like that if the story would still be completed this sprint, otherwise we should just make a better project-version selection widget an enhancement for later.

Can we separate the project-version selection part of the ui into its own component with some params so that we can use it in other places and have them all get enhancements for it? It may not be worthwhile yet since it should be easy to separate out later.